helix-core/src/main/java/org/apache/helix/controller/stages/CurrentStateComputationStage.java

```diff
@@ -191,6 +213,15 @@ private void updatePendingMessages(LiveInstance instance, Collection<Message> pe
     }
   }
 
+  private boolean isStaleMessage(Message message, String currentState) {
+    if (_isTaskFrameworkPipeline || currentState == null || message.getFromState() == null
```

Comment:
Introducing _isTaskFrameworkPipeline to isStaleMessage is confusing. I would prefer do this,
"if (!_isTaskFrameworkPipeline && isStaleMessage(...)) {set stale message}"